tests/trainer/optimization/test_manual_optimization.py

```diff
@@ -355,3 +357,271 @@ def configure_optimizers(self):
 
     num_manual_backward_calls = 3
     assert trainer.dev_debugger.count_events('backward_call') == limit_train_batches * num_manual_backward_calls
+
+
+class ExtendedModel(BoringModel):
+
+    count = 0
+    called = collections.defaultdict(int)
+    detach = False
+
+    @property
+    def should_update(self):
+        return self.count % 2 == 0
+
+    def on_train_batch_start(self, batch, batch_idx, dataloader_idx):
+        self.called["on_train_batch_start"] += 1
```

Comment:
maybe add another more general extension of boring model without would count calls and eventually dos some other debug duties?

Reply:
We will have to do so, but it is not really the scope of this PR.

Reply:
I would prefer for us to really think about creating a good testing framework for multi-nodes with all accelerators.

Reply:
> We will have to do so, but it is not really the scope of this PR.

yes, agree, shall we make some issue for it or where to collect these small notes?